sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -48,203 +47,231 @@ public class ImplementIDisposableCorrectly : SonarDiagnosticAnalyzer
 
         protected sealed override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSyntaxNodeActionInNonGenerated(CheckIDisposableImplementation, SyntaxKind.ClassDeclaration);
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+
+                var checker = new DisposableChecker(classDeclaration, c.SemanticModel);
+                var locations = checker.GetIssueLocations();
+                if (locations.Any())
+                {
+                    c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
+                        locations.ToAdditionalLocations(),
+                        locations.ToProperties()));
+                }
+            },
+            SyntaxKind.ClassDeclaration);
         }
 
-        private void CheckIDisposableImplementation(SyntaxNodeAnalysisContext c)
+        private class DisposableChecker
         {
-            var classDeclaration = (ClassDeclarationSyntax)c.Node;
-            var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+            private readonly ClassDeclarationSyntax classDeclaration;
+            private readonly SemanticModel semanticModel;
+            private readonly List<SecondaryLocation> secondaryLocations = new List<SecondaryLocation>();
+            private readonly INamedTypeSymbol classSymbol;
 
-            if (classSymbol == null || classSymbol.IsSealed)
+            public DisposableChecker(ClassDeclarationSyntax classDeclaration, SemanticModel semanticModel)
             {
-                return;
+                this.classDeclaration = classDeclaration;
+                this.semanticModel = semanticModel;
+                this.classSymbol = semanticModel.GetDeclaredSymbol(classDeclaration);
             }
 
-            var idisposableInterface = classSymbol.Interfaces.FirstOrDefault(IsOrImplementsIDisposable);
-            if (idisposableInterface == null)
+            public IEnumerable<SecondaryLocation> GetIssueLocations()
             {
-                return;
+                if (classSymbol == null || classSymbol.IsSealed)
+                {
+                    return Enumerable.Empty<SecondaryLocation>();
+                }
+
+                if (classSymbol.BaseType.Implements(KnownType.System_IDisposable))
+                {
+                    var idisposableInterfaceSyntax = classDeclaration.BaseList?.Types
+                        .FirstOrDefault(IsOrImplementsIDisposable);
+
+                    if (idisposableInterfaceSyntax != null)
+                    {
+                        AddSecondaryLocation(idisposableInterfaceSyntax.GetLocation(),
+                            $"Remove 'IDisposable' from the list of interfaces implemented by '{classSymbol.Name}'"
+                                + " and override the base class 'Dispose' implementation instead.");

```

Comment:
I know that I proposed that, but I don't think we should remove it in case we cannot find base.Dispose() because the base class could be exposing a method with different name and we could confuse the developer